### PR TITLE
Generic signature API

### PR DESCRIPTION
**Pull request by bwesterb**
Used to easily add support for new (PQ) signature schemes to [our fork of Go](https://github.com/cloudflare/go) and cfssl.

### CODE REVIEW

**Comment by bwesterb** [comment]
Another reason is that I'm quickly prototyping integration into Go stdlib to find possible issues with the API.  I found one already: we can't use `PubicKeyAlgorithm() x509.PublicKeyAlgorithm` in our interface for that would cause an import cycle.

**Comment by bwesterb** [comment]
@armfazh @claucece Have another look.

**Comment by claucece** [comment]
I'll check this tomorrow ;)

File: sign/api.go [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+		if tlsScheme.TLSIdentifier() == id {
+			return scheme
+		}
+	}
+	return nil
+}
+
+func UnmarshalPEMPublicKey(data []byte) (PublicKey, error) {
+	block, rest := pem.Decode(data)
+	if len(rest) != 0 {
```
Comment (armfazh) [inline sign/api.go:157]: I think the checks are for the block, like in the [Decode example](https://golang.org/pkg/encoding/pem/#Decode)

```
        block, rest := pem.Decode(pubPEMData)
	if block == nil || block.Type != "PUBLIC KEY" {
		log.Fatal("failed to decode PEM block containing public key")
	}
```

File: sign/eddilithium3.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	"github.com/cloudflare/circl/sign/eddilithium3"
+
+	cryptoRand "crypto/rand"
+	"encoding/asn1"
+	"errors"
+)
+
+type edDilithium3Scheme struct{}
+
+var EdDilithium3 Scheme = &edDilithium3Scheme{}
```
Comment (armfazh) [inline sign/eddilithium3.go:13]: it does not require the address

File: sign/eddilithium3.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+import (
+	"github.com/cloudflare/circl/sign/eddilithium3"
+
+	cryptoRand "crypto/rand"
+	"encoding/asn1"
+	"errors"
+)
+
+type edDilithium3Scheme struct{}
```
Comment (armfazh) [inline sign/eddilithium3.go:11]: All these definitions must live inside of `sign/eddilithium3` package,

Comment (armfazh) [inline sign/eddilithium3.go:11]: This is a zero-sized data, which is ok. What if this is an integer, which can also be used as identifier in the list of schemes.

**Comment by claucece** [comment]
I think it is looking good. Maybe we should be more clear on some names and all, though

File: sign/eddilithium3.go (hunk 2 above)
Comment (bwesterb) [reply sign/eddilithium3.go:13]: Try it without the `&`.

File: sign/eddilithium3.go (hunk 3 above)
Comment (bwesterb) [reply sign/eddilithium3.go:11]: That is what I would like as well, but how do you do that without creating an import cycle?

File: sign/api.go [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+	PublicKeySize() uint
+
+	// Size of binary marshalled public keys
+	PrivateKeySize() uint
+
+	// Name of the scheme
+	Name() string
+
+	// Size of signatures
+	SignatureSize() uint
```
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/api.go:85]: Sizes are always positive.  However, it might indeed be better to use `int` as it removed the need for casting, say, `len(something)` to `uint`.

File: sign/api.go [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+	// For compatibility with Go standard library
+	crypto.Signer
+
+	encoding.BinaryMarshaler
+}
+
+// A Scheme represents a specific instance of a signature scheme.
+type Scheme interface {
+	// GenerateKey creates a new key-pair.
+	GenerateKey() (PublicKey, PrivateKey, error)
```
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/api.go:52]: The common case should be as easy and foolproof as possible and the common case is that people don't  need to use a custom random source.

File: sign/api.go [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+}
+
+// SchemeByName returns the scheme with the given name and nil if it is not
+// supported.  Use ListSchemes() to list supported schemes.
+// Names are case insensitive.
+func SchemeByName(name string) Scheme {
+	// XXX add a (compile time?) lookup table
+	name = strings.ToLower(name)
+	for _, scheme := range schemes {
+		if strings.ToLower(scheme.Name()) == name {
```
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/api.go:111]: If we would remove this `ToLower()` then we will never match any scheme that has an upper case character in the return of its `scheme.Name()`.

File: sign/eddilithium3.go (hunk 3 above)
Comment (bwesterb) [reply sign/eddilithium3.go:11]: You can't use a specific integer as a type.

File: sign/api.go [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+		data,
+	}
+
+	return asn1.Marshal(pkix)
+}
+
+// Go1.15 adds {PublicKey,PrivateKey}.Equal(). Until then, we can use that
+// we use this.
+
+func PublicKeysEqual(a, b PublicKey) bool {
```
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/api.go:316]: I could use the interface that we expect Go1.15 to have, but I'd prefer to wait in case they change their mind.

File: sign/api.go [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+	for _, scheme := range schemes {
+		if strings.ToLower(scheme.Name()) == name {
+			return scheme
+		}
+	}
+	return nil
+}
+
+// ListSchemeNames returns the names of all schemes supported.
+func ListSchemeNames() []string {
```
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/api.go:119]: I'm thinking of adding such a function as well to return `[]Scheme`.

File: sign/eddilithium3.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+	if err != nil {
+		return nil, nil, err
+	}
+	return wrapPublicKey(pk, EdDilithium3), wrapPrivateKey(sk, EdDilithium3), nil
+}
+
+func (s *edDilithium3Scheme) Sign(sk PrivateKey, message []byte,
+	opts *SignatureOpts) []byte {
+	sig := make([]byte, eddilithium3.SignatureSize)
+	if opts != nil && opts.Context != "" {
```
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/eddilithium3.go:26]: I don't understand.  For `opts.Context` to make sense, we must have `opts != nil`.

**Comment by bwesterb** [comment]
> Maybe we should be more clear on some names and all, though

Sure, which?

File: sign/eddilithium3.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (armfazh) [reply sign/eddilithium3.go:26]: if any of the options is not used, then is ignored.

**Comment by armfazh** [comment]
I just added a generic interface for four signatures, there are some methods marked with TODO.
Also I moved some methods into the `circl/pki` package.

File: sign/api.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (claucece) [reply sign/api.go]: Yes, but perhaps the `scheme.Name()` should always be enforced to be lower case... what do you think?

File: sign/api.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (claucece) [reply sign/api.go]: I agree now.

File: sign/api.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (claucece) [reply sign/api.go]: Yeah.. I was thinking from the casting perspective that it is used many times...

**Comment by claucece** [comment]
> Sure, which?

Let me thinking of a nice naming convention and propose something over here tomorrow.

File: sign/api.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/api.go]: Ok, let's go for int!

File: sign/ed25519/signapi.go [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+	"github.com/cloudflare/circl/sign"
+)
+
+// Scheme is
+const Scheme = scheme(sign.Ed25519)
+
+type scheme sign.SchemeID
+
+func (scheme) ID() sign.SchemeID    { return sign.SchemeID(Scheme) }
```
Comment (bwesterb) [inline sign/ed25519/signapi.go:15]: What is the advantage of this compared to:

```
type scheme struct{}
func (scheme) ID() sign.SchemeID    { return sign.Ed25519 }
```

File: sign/api/api.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+var allSchemes [sign.SchemeCount]sign.Scheme
+var allSchemeNames map[string]sign.Scheme
+
+func init() {
+	allSchemeNames = make(map[string]sign.Scheme)
+	register(ed25519.Scheme)
+	register(ed448.Scheme)
+	register(eddilithium3.Scheme)
+	register(eddilithium4.Scheme)
```
Comment (bwesterb) [inline sign/api/api.go:20]: Why not `register(sign.Ed25519, ed25519.Scheme)`?  Then there is no need for a public `ID()`, `SchemeID`, etc.

File: sign/api/api.go [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+	allSchemes[s.ID()] = s
+	allSchemeNames[s.Name()] = s
+}
+
+// SchemeByName returns the scheme with the given name and nil if it is not
+// supported.
+func SchemeByName(name string) sign.Scheme { return allSchemeNames[name] }
+
+// AllSchemes returns all signature schemes supported.
+func AllSchemes() []sign.Scheme { a := allSchemes; return a[:] }
```
Comment (bwesterb) [inline sign/api/api.go:33]: This allows the caller to manipulate the list. That's dangerous.

**Comment by bwesterb** [comment]
I have a few issues with your changes @armfazh 

- I dislike that `SchemeByName` is now under the `github.com/cloudflare/circl/sign/api` package.
- Providing a context to a scheme that does not support it is completely ignored.  It should be an error.

File: pki/pki.go [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+	"github.com/cloudflare/circl/sign"
+	"github.com/cloudflare/circl/sign/api"
+)
+
+var allSchemesByOID map[string]sign.Scheme
+var allSchemesByTLS map[uint]sign.Scheme
+
+func init() {
+	allSchemesByOID = make(map[string]sign.Scheme)
+	for _, scheme := range api.AllSchemes() {
```
Comment (bwesterb) [inline pki/pki.go:19]: @armfazh You can't rely on the right `init()` order.

Comment (bwesterb) [reply pki/pki.go:19]: https://medium.com/@markbates/go-init-order-dafa89fcef22

File: sign/api/api.go (hunk 12 above)
Comment (armfazh) [reply sign/api/api.go:33]: `a` is a copy, not a reference to the register

Comment (bwesterb) [reply sign/api/api.go:33]: Yes, you're right. I didn't read it correctly.

File: pki/pki.go (hunk 13 above)
Comment (mmcloughlin) [reply pki/pki.go:19]: > If a package has imports, the imported packages are initialized before initializing the package itself.

https://golang.org/ref/spec#Package_initialization

File: pki/pki_test.go [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+import (
+	"testing"
+
+	"github.com/cloudflare/circl/pki"
+	"github.com/cloudflare/circl/sign/api"
+)
+
+func TestPEM(t *testing.T) {
+	for _, scheme := range api.AllSchemes() {
```
Comment (mmcloughlin) [inline pki/pki_test.go:11]: This seems like an appropriate use for subtests.

```go
func TestPEM(t *testing.T) {
	for _, scheme := range api.AllSchemes() {
		t.Run(scheme.Name(), func(t *testing.T) {
			// ...
		}
	}
}
```

File: sign/api/api_test.go [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+	"fmt"
+	"testing"
+
+	"github.com/cloudflare/circl/sign"
+	"github.com/cloudflare/circl/sign/api"
+)
+
+func TestApi(t *testing.T) {
+	allSchemes := api.AllSchemes()
+	for _, scheme := range allSchemes {
```
Comment (mmcloughlin) [inline sign/api/api_test.go:14]: Consider subtests?

File: sign/ed25519/ed25519_test.go [hunk 16]
```diff
@@ -39,6 +39,24 @@ func TestMalleability(t *testing.T) {
...
+func TestPublic(t *testing.T) {
+	var zero zeroReader
+	pub, priv, err := ed25519.GenerateKey(zero)
+	if !priv.Equal(priv) {
+		t.Fatal()
+	}
+	if !pub.Equal(pub) {
+		t.Fatal()
+	}
+	if err != nil {
```
Comment (mmcloughlin) [inline sign/ed25519/ed25519_test.go:51]: Why is the error check not the first assertion?

File: sign/ed25519/ed25519_test.go [hunk 17]
```diff
@@ -39,6 +39,24 @@ func TestMalleability(t *testing.T) {
...
 }
 
+func TestPublic(t *testing.T) {
+	var zero zeroReader
+	pub, priv, err := ed25519.GenerateKey(zero)
+	if !priv.Equal(priv) {
+		t.Fatal()
+	}
+	if !pub.Equal(pub) {
+		t.Fatal()
```
Comment (mmcloughlin) [inline sign/ed25519/ed25519_test.go:49]: Why no message with `Fatal`? If you intend to use no message, just use `FailNow()` directly.

https://golang.org/pkg/testing/#T.Fatal

File: sign/ed25519/signapi.go [hunk 18]
```diff
@@ -0,0 +1,75 @@
+package ed25519
```
Comment (mmcloughlin) [inline sign/ed25519/signapi.go:1]: Nearly identical to the other `signapi.go` files (I know there are some differences). Is it generated? If so it needs a generated comment.

File: sign/ed25519/signapi.go [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+func (scheme) GenerateKey() (sign.PublicKey, sign.PrivateKey, error) {
+	return GenerateKey(rand.Reader)
+}
+
+func (scheme) Sign(
+	sk sign.PrivateKey,
+	message []byte,
+	opts *sign.SignatureOpts) []byte {
```
Comment (mmcloughlin) [inline sign/ed25519/signapi.go:33]: nit: my preference would be to use a trailing comma in multi-line signatures.

File: sign/sign.go [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+
+	// Size of signatures
+	SignatureSize() int
+
+	// Size of seeds
+	SeedSize() int
+}
+
+var (
+	// ErrType is
```
Comment (mmcloughlin) [inline sign/sign.go:99]: FYI, if you don't want to document these individually, I _think_ the linter accepts just a comment on the entire `var` block.

File: sign/sign.go [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+	ErrType = errors.New("types mismatch")
```
Comment (mmcloughlin) [inline sign/sign.go:100]: nit: `ErrTypeMismatch`?

**Comment by bwesterb** [comment]
@armfazh Will you make changes according to the comments or shall I make them?

**Comment by armfazh** [comment]
> @armfazh Will you make changes according to the comments or shall I make them?

Please, go forward with changes.

File: sign/api/api.go (hunk 11 above)
Comment (bwesterb) [reply sign/api/api.go]: @armfazh I removed `SchemeID` without changing the rest of the API at all.

File: pki/pki_test.go (hunk 14 above)
Comment (bwesterb) [reply pki/pki_test.go:11]: Nice, didn't know about these.  Thanks!

File: sign/ed25519/ed25519_test.go (hunk 17 above)
Comment (bwesterb) [reply sign/ed25519/ed25519_test.go:49]: Does it really matter?  (Changed anyway.)

File: sign/ed25519/ed25519_test.go (hunk 16 above)
Comment (bwesterb) [reply sign/ed25519/ed25519_test.go:51]: That's a bug. Thanks for pointing that out.

File: sign/ed25519/signapi.go (hunk 18 above)
Comment (bwesterb) [reply sign/ed25519/signapi.go:1]: No, it's not.

File: sign/ed25519/signapi.go (hunk 19 above)
Comment (bwesterb) [reply sign/ed25519/signapi.go:31]: Yes, that's nicer.

**Comment by bwesterb** [comment]
@armfazh 

> * I dislike that `SchemeByName` is now under the `github.com/cloudflare/circl/sign/api` package.

I don't see a way to solve this without making something else ugly.

> * Providing a context to a scheme that does not support it is completely ignored.  It should be an error.

I've changed that behaviour: provided a context where none is supported causes an error.

**Comment by bwesterb** [comment]
I've also removed the `Hash` from the SignatureOpts as it's not used.

File: sign/api.go [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+		&pem.Block{
+			Type:  fmt.Sprintf("%s PRIVATE KEY", sk.Scheme().Name()),
+			Bytes: data,
+		},
+	)
+	return str, nil
+}
+
+func UnmarshalPKIXPrivateKey(data []byte) (PrivateKey, error) {
+	var pkix struct {
```
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/api.go]: For the private keys it is (and so I made a common type) — for the public keys it isn't.

File: sign/api.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/api.go]: `api.AllSchemes()`

File: sign/api.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/api.go]: @armfazh Removed case insensitivity again. Did you do that on purpose @armfazh ?

File: sign/eddilithium3.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (bwesterb) [reply sign/eddilithium3.go]: I don't understand — could you rephrase perhaps?

**Comment by bwesterb** [comment]
@armfazh @claucece I've addressed all comments.  Before starting with final polish (i.e. better documentation & examples), I think we still need to discuss some API issues:

- [ ] Are we really happy with having `sign` package which defines `Scheme` and `sign/api` package which actually lists them?  Isn't `schemes` perhaps a better name than `api` as the `sign` package itself defines the API.  Then `schemes.All()` and `scheme.ByName()` would be nice to write.
- [ ] Do we want Scheme names to be case insensitive?  This was the case but @armfazh reversed it. This is required for easy integration with `cfssl`.
- [ ] Doesn't ed25519 have a variant that support contexts?  Should we add this to the Ed25519 scheme or have a separate Ed25519Ctx scheme?  I guess it's best to have separate schemes for these variants — @claucece what do you think?

**Comment by armfazh** [comment]
> * [ ]  Are we really happy with having `sign` package which defines `Scheme` and `sign/api` package which actually lists them?  Isn't `schemes` perhaps a better name than `api` as the `sign` package itself defines the API.  Then `schemes.All()` and `scheme.ByName()` would be nice to write.

Agree on `s/api/schemes/` 

> * [ ]  Do we want Scheme names to be case insensitive?  This was the case but @armfazh reversed it. This is required for easy integration with `cfssl`.

I don't have a strong opinion on this. Feel free to bring back that code.

> * [ ]  Doesn't ed25519 have a variant that support contexts?  Should we add this to the Ed25519 scheme or have a separate Ed25519Ctx scheme?  I guess it's best to have separate schemes for these variants — @claucece what do you think?

That variant would be another named scheme `Ed25519Ctx`, which is separated from the current `Ed25519` scheme.

> I've also removed the Hash from the SignatureOpts as it's not used — we can add it back in if we have a scheme that actually uses it.

I added that the Hash field to make the structure compatible with the crypto.SignerOpts. So, we can receive a crypto.SignerOpts, and internally convert it to a different type.

> I dislike that SchemeByName is now under the github.com/cloudflare/circl/sign/api package.

I also didn't like it at all. But, that's the way I found for avoiding circle dependencies. Renaming sign/api to sign/schemes can alleviate a bit its usage.

File: sign/ed25519/signapi.go [hunk 23]
```diff
@@ -0,0 +1,84 @@
+package ed25519
+
+import (
+	"crypto/rand"
+	"encoding/asn1"
+
+	"github.com/cloudflare/circl/sign"
+)
+
+var Scheme sign.Scheme = &scheme{}
```
Comment (armfazh) [inline sign/ed25519/signapi.go:10]: I don't understand why to take the address of a zero-sized value. You can implement the interface with `scheme` and not with `*scheme`.

File: sign/sign.go [hunk 24]
```diff
@@ -0,0 +1,103 @@
...
+	PrivateKeySize() int
+
+	// Size of signatures
+	SignatureSize() int
+
+	// Size of seeds
+	SeedSize() int
+
+	// Returns whether contexts are supported
+	SupportsContext() bool
```
Comment (armfazh) [inline sign/sign.go:80]: I am not sure whether this method is required. As per the policy mentioned above, if context is provided and not required then it should fail.

Comment (bwesterb) [reply sign/sign.go:80]: It's nice to know beforehand whether a scheme will by useable when you require a context.  (I need this in the tests to check whether to provide a context or not.)

Comment (armfazh) [reply sign/sign.go:80]: Is this the only use case? I think this issue was one of the reasons of the previous policy. A scheme uses the context if the scheme supports it, otherwise it is simply ignored. 
Anyway, either policy works.

Comment (bwesterb) [reply sign/sign.go:80]: Ignoring context if the scheme doesn't support it is very dangerous: a protocol might essentially require different signatures for different contexts.  If we ignore the context for schemes that don't support it, we leave a foot gun.

File: sign/ed25519/signapi.go (hunk 23 above)
Comment (bwesterb) [reply sign/ed25519/signapi.go:10]: Under the hood an instance of an interface is always a pair of a pointer and a concrete type, so instantiating it with a pointer is closer to what really happens.  The assembly generated btw is exactly [the same](https://godbolt.org/z/6zq31x).

Comment (armfazh) [reply sign/ed25519/signapi.go:10]: The difference is on the assembly calls.
     `call  AX` vs `call    "".(*Impl2).A(SB)`
one call is with a register and the other is with a pointer. 

I am not convinced why to use pointer when they are not needed.

Comment (bwesterb) [reply sign/ed25519/signapi.go:10]: Oops, I simplified my example too much.

[Take a look at this one](https://godbolt.org/z/KsaccE)

Creating an instance of an interface with either a empty struct or a pointer to an empty struct results in the pointer `runtime.zerobase`.  (The code of `alloc1` and `alloc2` is the same.  This time I checked properly.)

The generic way to run a function on an interface involves a virtual call (`call ax`).  So interestingly the Go compiler optimised the call with the pointer receiver, but not the call with the value receiver.  I don't see a reason why it couldn't optimise the one with the value receiver as well and perhaps it will in the future.

So using pointers or not does not actually change anything when using empty structs (and the calling code doesn't know the types.) And if it does it optimizes the pointer case better. That's not the reason I use pointers though — as I said I prefer putting them here as it's closer to what Go actually does under the hood: an instance of an interface is always a pointer.

**Comment by bwesterb** [comment]
> Agree on `s/api/schemes/`

Done.  Updated this branch, Go fork and cfssl fork.

> I added that the Hash field to make the structure compatible with the crypto.SignerOpts. So, we can receive a crypto.SignerOpts, and internally convert it to a different type.

The type of `Scheme.Sign` is `Sign(sk PrivateKey, message []byte, opts *SignatureOpts) []byte` so that wouldn't accept a `crypto.SignerOpts` anyway now.  (Note that there are two sign functions: one on the scheme and one on the PrivateKey — the one on the private key does accept a `crypto.SignerOpts` but does not allow for the advanced options like `Context`.)

**Comment by armfazh** [comment]
> the one on the private key does accept a crypto.SignerOpts but does not allow for the advanced options like Context.)

I had the same intuition, but here is the trick. Even if we announce to receive a `crypto.SignerOpts`, we could pass a `Sign.SignatureOpts` (which must have field Hash) and internally convert the parameter into a `Sign.SignatureOpts` struct. (I learn this trick from @claucece)

File: sign/sign.go (hunk 24 above)
Comment (armfazh) [reply sign/sign.go:84]: good point.

File: sign/ed25519/signapi.go (hunk 23 above)
Comment (armfazh) [reply sign/ed25519/signapi.go:10]: > So interestingly the Go compiler optimised the call with the pointer receiver, but not the call with the value receiver.

I want to learn more about this.

> So using pointers or not does not actually change anything when using empty structs.

Agree.

**Comment by claucece** [comment]
>  Are we really happy with having sign package which defines Scheme and sign/api package which actually lists them? Isn't schemes perhaps a better name than api as the sign package itself defines the API. Then schemes.All() and scheme.ByName() would be nice to write.

Yes! 

> Do we want Scheme names to be case insensitive? This was the case but @armfazh reversed it. This is required for easy integration with cfssl.

Mmm... It will be better case insensitive.

> Doesn't ed25519 have a variant that support contexts? Should we add this to the Ed25519 scheme or have a separate Ed25519Ctx scheme? I guess it's best to have separate schemes for these variants — @claucece what do you think?

There are two: the ctx and the ph. The schemes should be there on the ed25519 API.

**Comment by claucece** [comment]
> I had the same intuition, but here is the trick. Even if we announce to receive a crypto.SignerOpts, we could pass a Sign.SignatureOpts (which must have field Hash) and internally convert the parameter into a Sign.SignatureOpts struct. (I learn this trick from @claucece)

Exactly! Then, the "Golang-way" API does not change ;)

**Comment by claucece** [comment]
@bwesterb check what I did with the ed448 and ed25519 API, specially the usage of this: https://github.com/cloudflare/circl/blob/master/sign/ed25519/ed25519.go#L69

File: sign/ed25519/ed25519_test.go (hunk 17 above)
Comment (bwesterb) [reply sign/ed25519/ed25519_test.go]: It does matter: FailNow doesn't show the line number when triggered and Fatal does.  Thus I reverted back to using Fatal.

**Comment by bwesterb** [comment]
> I had the same intuition, but here is the trick. Even if we announce to receive a `crypto.SignerOpts`, we could pass a `Sign.SignatureOpts` (which must have field Hash) and internally convert the parameter into a `Sign.SignatureOpts` struct. (I learn this trick from @claucece)

Yes, that's for a method on the private key, but here we're dealing with a top level function.

(For the `Sign` method on the private key it's actually a bit tricky.  If we use the same type for an ed25519 and ed25519ctx private key, then we can't know whether it was used for the scheme ed25519 or ed25519ctx.  I'd like to propose to fix this in a coming pull request where we add schemes for the variants of ed*.)

**Comment by bwesterb** [comment]
@armfazh @claucece I'm happy with the PR as it is now (assuming we'll be able to make changes to the API before we tag it.) Do you agree?

**Comment by claucece** [comment]
> (For the Sign method on the private key it's actually a bit tricky. If we use the same type for an ed25519 and ed25519ctx private key, then we can't know whether it was used for the scheme ed25519 or ed25519ctx. I'd like to propose to fix this in a coming pull request where we add schemes for the variants of ed*.)

Mmm.. ok.

> Do you agree?

Sure!

File: pki/pki.go [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+	})
+}
+
+func MarshalPEMPrivateKey(sk sign.PrivateKey) ([]byte, error) {
+	data, err := MarshalPKIXPrivateKey(sk)
+	if err != nil {
+		return nil, err
+	}
+	str := pem.EncodeToMemory(&pem.Block{
+		Type:  fmt.Sprintf("%s PRIVATE KEY", sk.Scheme().Name()),
```
Comment (armfazh) [inline pki/pki.go:158]:
```suggestion
		Type:  sk.Scheme().Name() + " PRIVATE KEY",
```

Comment (armfazh) [reply pki/pki.go:158]: During unmarshalling, does this name is verified that matches with the content?

File: pki/pki.go [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+	return asn1.Marshal(struct {
+		pkix.AlgorithmIdentifier
+		asn1.BitString
+	}{
+		pkix.AlgorithmIdentifier{
+			Algorithm: scheme.(CertificateScheme).Oid(),
+		},
+		asn1.BitString{
+			Bytes:     data,
+			BitLength: len(data) * 8,
```
Comment (armfazh) [inline pki/pki.go:147]: Do we need to verify that`scheme.PublicKeySize() == len(data)` ?

File: pki/pki.go [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+func init() {
+	allSchemesByOID = make(map[string]sign.Scheme)
+	for _, scheme := range schemes.All() {
+		if cert, ok := scheme.(CertificateScheme); ok {
+			allSchemesByOID[cert.Oid().String()] = scheme
+		}
+	}
+
+	allSchemesByTLS = make(map[uint]sign.Scheme)
+	for _, scheme := range schemes.All() {
```
Comment (armfazh) [inline pki/pki.go:33]: Merge for-loops.

File: pki/pki.go (hunk 25 above)
Comment (bwesterb) [reply pki/pki.go]: > During unmarshalling, does this name is verified that matches with the content?

No.  The reason is that I'm not sure that all private key files actually contain the type in them.

File: pki/pki.go (hunk 26 above)
Comment (bwesterb) [reply pki/pki.go:147]: I don't think it's necessary here.

**Comment by bwesterb** [comment]
Thank you, I merged it.  I created an issue ( #147 ) to track desired improvements to this API.
